File: Assignment1/q2/q2.py
from PIL import Image

# ================================
# NODE DEFINITION
# ================================
class Node:
    def __init__(self, i, j):
        self.right = None
        self.left = None
        self.up = None
        self.down = None
        self.next = None
        self.cell_position = (i, j)


# ================= READ PNG =================
def readPNG(filename):
    img = Image.open(filename).convert("L")  # grayscale
    width, height = img.size
    pixels = img.load()

    matrix = [[0]*width for _ in range(height)]
    for i in range(height):
        for j in range(width):
            matrix[i][j] = 1 if pixels[j, i] < 128 else 0

    return matrix, width, height


# ================= WRITE PNG =================
def writePNG(filename, matrix):
    height = len(matrix)
    width = len(matrix[0])

    img = Image.new("L", (width, height))
    pixels = img.load()

    for i in range(height):
        for j in range(width):
            pixels[j, i] = 0 if matrix[i][j] else 255

    img.save(filename)


# ================= SPARSE GRAPH =================
def convertToLinkedList(matrix):
    m, n = len(matrix), len(matrix[0])
    node_matrix = [[None]*n for _ in range(m)]

    head = tail = None

    for i in range(m):
        for j in range(n):
            if matrix[i][j] == 1:
                node = Node(i, j)
                node_matrix[i][j] = node
                if not head:
                    head = tail = node
                else:
                    tail.next = node
                    tail = node

    dx = [1, -1, 0, 0]
    dy = [0, 0, 1, -1]

    for i in range(m):
        for j in range(n):
            if not node_matrix[i][j]:
                continue
            for k in range(4):
                x, y = i + dx[k], j + dy[k]
                if 0 <= x < m and 0 <= y < n and node_matrix[x][y]:
                    if dx[k] == 1:
                        node_matrix[i][j].down = node_matrix[x][y]
                    if dx[k] == -1:
                        node_matrix[i][j].up = node_matrix[x][y]
                    if dy[k] == 1:
                        node_matrix[i][j].right = node_matrix[x][y]
                    if dy[k] == -1:
                        node_matrix[i][j].left = node_matrix[x][y]

    return head


# ================= FLOOD FILL (NO DFS) =================
def floodFillComponent(start_node, visited):
    component_nodes = set()
    component_nodes.add(start_node)

    changed = True
    while changed:
        changed = False
        for node in list(component_nodes):
            for nbr in [node.up, node.down, node.left, node.right]:
                if nbr and nbr not in component_nodes:
                    component_nodes.add(nbr)
                    changed = True

    # mark visited
    for node in component_nodes:
        visited.add(node)

    # compute area & boundary
    area = len(component_nodes)
    boundary = []

    for node in component_nodes:
        for nbr in [node.up, node.down, node.left, node.right]:
            if nbr is None:
                boundary.append(node)
                break

    return area, boundary


# ================= CONNECTED COMPONENTS =================
def findConnectedComponents(head):
    visited = set()
    component = 0

    curr = head
    while curr:
        if curr not in visited:
            component += 1
            area, boundary = floodFillComponent(curr, visited)

            print(f"Object {component}")
            print(f"Area: {area}")
            print("Boundary pixels:", end=" ")
            for b in boundary:
                print(b.cell_position, end=" ")
            print("\n")

        curr = curr.next

    print("Total objects detected =", component)


# ================= FLIP IMAGE =================
def flipImage(matrix):
    h, w = len(matrix), len(matrix[0])
    flipped = [[0]*w for _ in range(h)]

    for i in range(h):
        for j in range(w):
            flipped[i][j] = 0 if matrix[i][j] else 1

    return flipped


# ================= MAIN =================
if __name__ == "__main__":
    inputFile = "input_image.png"
    outputFile = "flipped_image.png"

    matrix, width, height = readPNG(inputFile)
    head = convertToLinkedList(matrix)

    findConnectedComponents(head)

    flipped = flipImage(matrix)
    writePNG(outputFile, flipped)

    print("Flipped image saved as", outputFile)

File: Assignment1/q2/q2.cpp
#include <bits/stdc++.h>
using namespace std;

struct Node {
    Node* right;
    Node* left;
    Node* up;
    Node* down;
    Node* next;

    pair<int,int> cell_position;

    Node(int i, int j) {
        right = left = up = down = next = nullptr;
        cell_position = {i, j};
    }
};

/* ================= READ PGM ================= */
vector<vector<int>> readPGM(const string& filename, int& width, int& height, int& maxval) {
    ifstream file(filename, ios::binary);
    if (!file) {
        cerr << "Cannot open image file\n";
        exit(1);
    }

    string magic;
    file >> magic;
    if (magic != "P2" && magic != "P5") {
        cerr << "Unsupported PGM format\n";
        exit(1);
    }

    file >> width >> height >> maxval;
    file.ignore();

    vector<vector<int>> image(height, vector<int>(width, 0));

    if (magic == "P2") {
        for (int i = 0; i < height; i++)
            for (int j = 0; j < width; j++) {
                int pixel;
                file >> pixel;
                image[i][j] = (pixel == maxval ? 0 : 1);
            }
    } else { // P5
        for (int i = 0; i < height; i++)
            for (int j = 0; j < width; j++) {
                unsigned char pixel;
                file.read(reinterpret_cast<char*>(&pixel), 1);
                image[i][j] = (pixel == maxval ? 0 : 1);
            }
    }

    return image;
}

/* ================= WRITE PGM ================= */
void writePGM(const string& filename,
              const vector<vector<int>>& image,
              int width, int height, int maxval) {

    ofstream file(filename, ios::binary);
    if (!file) {
        cerr << "Cannot write output image\n";
        exit(1);
    }

    file << "P2\n";
    file << width << " " << height << "\n";
    file << maxval << "\n";

    for (int i = 0; i < height; i++) {
        for (int j = 0; j < width; j++) {
            file << (image[i][j] ? maxval : 0) << " ";
        }
        file << "\n";
    }
}

/* ================= SPARSE GRAPH ================= */
Node* convertToLinkedList(vector<vector<int>>& matrix,
                          vector<vector<Node*>>& node_matrix) {

    int m = matrix.size();
    int n = matrix[0].size();

    node_matrix.assign(m, vector<Node*>(n, nullptr));

    Node* head = nullptr;
    Node* tail = nullptr;

    for (int i = 0; i < m; i++) {
        for (int j = 0; j < n; j++) {
            if (matrix[i][j] == 1) {
                Node* node = new Node(i, j);
                node_matrix[i][j] = node;

                if (!head) head = tail = node;
                else {
                    tail->next = node;
                    tail = node;
                }
            }
        }
    }

    int dx[4] = {1, -1, 0, 0};
    int dy[4] = {0, 0, 1, -1};

    for (int i = 0; i < m; i++) {
        for (int j = 0; j < n; j++) {
            if (!node_matrix[i][j]) continue;

            for (int k = 0; k < 4; k++) {
                int x = i + dx[k];
                int y = j + dy[k];

                if (x >= 0 && y >= 0 && x < m && y < n &&
                    node_matrix[x][y]) {

                    if (dx[k] == 1)  node_matrix[i][j]->down  = node_matrix[x][y];
                    if (dx[k] == -1) node_matrix[i][j]->up    = node_matrix[x][y];
                    if (dy[k] == 1)  node_matrix[i][j]->right = node_matrix[x][y];
                    if (dy[k] == -1) node_matrix[i][j]->left  = node_matrix[x][y];
                }
            }
        }
    }

    return head;
}

/* ================= CONNECTED COMPONENTS ================= */
void dfs(Node* node,
         unordered_set<Node*>& visited,
         int& area,
         vector<Node*>& boundary) {

    visited.insert(node);
    area++;

    bool isBoundary = false;

    Node* neighbors[4] = {
        node->up, node->down, node->left, node->right
    };

    for (Node* nbr : neighbors) {
        if (!nbr) isBoundary = true;
        else if (!visited.count(nbr))
            dfs(nbr, visited, area, boundary);
    }

    if (isBoundary)
        boundary.push_back(node);
}

void findConnectedComponents(Node* head) {
    unordered_set<Node*> visited;
    int component = 0;

    for (Node* curr = head; curr; curr = curr->next) {
        if (!visited.count(curr)) {
            component++;
            int area = 0;
            vector<Node*> boundary;

            dfs(curr, visited, area, boundary);

            cout << "Object " << component << "\n";
            cout << "Area: " << area << "\n\n";

            cout << "Boundary pixels: ";
            for (auto* b : boundary) {
                cout << "(" << b->cell_position.first
                    << "," << b->cell_position.second << ") ";
            }
            cout << "\n\n";

        }
    }

    cout << "Total objects detected = " << component << "\n";
}

/* ================= FLIP IMAGE ================= */
vector<vector<int>> flipImage(const vector<vector<int>>& original) {
    int h = original.size();
    int w = original[0].size();

    vector<vector<int>> flipped(h, vector<int>(w, 0));
    for (int i = 0; i < h; i++)
        for (int j = 0; j < w; j++)
            flipped[i][j] = original[i][j] ? 1 : 0;

    return flipped;
}

/* ================= MAIN ================= */
int main() {
    string inputFile = "input_image.pgm";
    string outputFile = "flipped_image.pgm";

    int width, height, maxval;
    vector<vector<int>> matrix = readPGM(inputFile, width, height, maxval);

    vector<vector<Node*>> node_matrix;
    Node* head = convertToLinkedList(matrix, node_matrix);

    findConnectedComponents(head);

    vector<vector<int>> flipped = flipImage(matrix);
    writePGM(outputFile, flipped, width, height, maxval);

    cout << "Flipped image saved as " << outputFile << "\n";

    return 0;
}
